File: backend/src/models/comments.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

/// Represents a comment returned by the comments API.
#[derive(Debug, Serialize, sqlx::FromRow)]
#[serde(rename_all = "camelCase")]
pub struct Comment {
    pub id: Uuid,
    pub post_id: Uuid,
    pub author_id: Uuid,
    pub parent_comment_id: Option<Uuid>,
    pub body: String,
    pub status: String,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
    pub deleted_at: Option<DateTime<Utc>>,
}

/// Accepts the body and optional parent of a new comment.
#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateCommentRequest {
    pub body: String,
    pub parent_comment_id: Option<Uuid>,
}

/// Accepts a replacement body for an existing comment.
#[derive(Debug, Deserialize)]
pub struct UpdateCommentRequest {
    pub body: String,
}

#[cfg(test)]
#[path = "../../tests/unit/models/comments.rs"]
mod tests;

File: blog/src/lib/models/comments.ts
/** Represents a post comment returned by the backend. */
export interface Comment {
	id: string;
	postId: string;
	authorId: string;
	parentCommentId: string | null;
	body: string;
	status: string;
	createdAt: string;
	updatedAt: string;
	deletedAt: string | null;
}

/** Supplies the mutable fields for a comment update. */
export interface UpdateCommentInput {
	body: string;
}

/** Supplies the content and optional parent for a new comment. */
export interface CreateCommentInput {
	body: string;
	parentCommentId?: string | null;
}
